aesd-char-driver/main.c

```diff
@@ -0,0 +1,286 @@
+/**
+ * @file aesdchar.c
+ * @brief Functions and data related to the AESD char driver implementation
+ *
+ * Based on the implementation of the "scull" device driver, found in
+ * Linux Device Drivers example code.
+ *
+ * @author Dan Walkes
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/syscalls.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/cdev.h>
+#include <linux/fs.h> // file_operations
+#include <asm/uaccess.h>
+#include "aesdchar.h"
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("Jordi Cros Mompart");
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    //Handle open
+    PDEBUG("open\n");
+    try_module_get(THIS_MODULE);
+
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    //Handle release
+    PDEBUG("release\n");
+    module_put(THIS_MODULE);
+
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    size_t  entry_offset = 0;
+    struct aesd_buffer_entry *entry = NULL;
+    PDEBUG("read %zu bytes with offset %lld\n",count,*f_pos);
+    
+    //"count" is the number of bytes to be returned
+    //"f_pos" is the offset from where to start the read
+
+    //Set mutex
+    mutex_lock(&aesd_device.lock);
```

Comment:
@jordicrosmompart I suggest the interruptible version here to support exit on Ctrl->C or signals, will need to check the return code in that case.